sdk/samples/iot/pnp/pnp_thermostat_component.c

```diff
@@ -22,6 +21,7 @@
 #include <azure/core/az_json.h>
 #include <azure/core/az_result.h>
 #include <azure/core/az_span.h>
+#include <azure/iot/az_iot_common.h>
```

Comment:
What made you decide to explicitly include this header, when it isn't _technically_ necessary (because other headers are transitively including it anyway)?

I think it is a good idea, but I wonder what other such headers are missing. How would we detect that vs. figure out when a header is unused and need to remove it?

Reply:
Hi.  Added comment below.  Yes, there is more onus on the developer to stay up-to-date on what headers to remove and add since the compiler will not automatically detect it.  It is not critical, but I think a good practice to keep headers up to date when one is going to add standard functionality that would require a header (even if it is already included transitively).

Reply:
> It is not critical, but I think a good practice to keep headers up to date when one is going to add standard functionality that would require a header (even if it is already included transitively).

Let's make sure we are discussing the same thing to avoid conflating audiences.

One topic is on what we want **_customers of the SDK_** to do, which is the primarily purpose of the samples. Your comment below is talking about that (and using single headers vs. many).

My comment here, is about the **author of samples and contributors to this repo and SDK itself**. For such developers, I understand and agree that we want to use granular header files, only include the ones we need, and include them explicitly.

My question is, how do we make sure that happens within our repo? This is not something we can detect by code reviews without investing a lot of time. So, is there a compiler tool/static analysis that can help make sure all transitive header files are explicitly included, and also one which highlights unused headers or dead code paths? Even if we thoroughly go through the SDK and fix all such instances today, how do we plan to stay green? We need some sort of automation here. Got any ideas worth exploring?

Reply:
https://github.com/Azure/azure-sdk-for-c/issues/1371